controllers/managedcluster_controller.go

```diff
@@ -0,0 +1,131 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+
+	"github.com/red-hat-storage/odf-multicluster-orchestrator/controllers/utils"
+	viewv1beta1 "github.com/stolostron/multicloud-operators-foundation/pkg/apis/view/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ManagedClusterReconciler struct {
+	Client client.Client
+	Logger *slog.Logger
+}
+
+const (
+	OdfInfoClusterClaimNamespacedName = "odfinfo.odf.openshift.io"
+)
+
+func (r *ManagedClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.Logger.With("ManagedCluster", req.NamespacedName)
+	logger.Info("Reconciling ManagedCluster")
+
+	var managedCluster clusterv1.ManagedCluster
+	if err := r.Client.Get(ctx, req.NamespacedName, &managedCluster); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error("Failed to get ManagedCluster", "error", err)
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err := r.processManagedClusterViews(ctx, managedCluster); err != nil {
+		logger.Error("Failed to ensure ManagedClusterView", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Successfully reconciled ManagedCluster")
+
+	return ctrl.Result{}, nil
+}
+
+func hasRequiredODFKey(mc *clusterv1.ManagedCluster) bool {
+	claims := mc.Status.ClusterClaims
+	for _, claim := range claims {
+		if claim.Name == OdfInfoClusterClaimNamespacedName {
+			return true
+		}
+	}
+	return false
+
+}
+func (r *ManagedClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.Logger.Info("Setting up ManagedClusterReconciler with manager")
+	managedClusterPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			obj, ok := e.ObjectNew.(*clusterv1.ManagedCluster)
+			if !ok {
+				return false
+			}
+			return hasRequiredODFKey(obj)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			obj, ok := e.Object.(*clusterv1.ManagedCluster)
+			if !ok {
+				return false
+			}
+			return hasRequiredODFKey(obj)
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.ManagedCluster{}, builder.WithPredicates(managedClusterPredicate, predicate.GenerationChangedPredicate{})).
```

Comment:
Do we need both predicates? We already have a custom Predicate where we are checking for create and update event of ManagedClusters with ClusterClaim.